# Phase 2: Machine Learning Feature Engineering & Training Pipeline

## Executive Summary & Pipeline Workflow

This notebook transforms preprocessed 1-minute cryptocurrency k-lines into a predictive machine learning pipeline. It engineers 16 domain-specific features, implements strict chronological data splitting with label window purging to prevent data leakage, and trains a benchmark ladder of 5 modeling approaches:

1. **Majority Class Floor Baseline**: Predicts the most common market direction (constant baseline).
2. **Continuous OLS Return Regression**: Fits linear regression on 15-minute future returns and thresholds the forecast.
3. **Logistic Regression Classifier**: Interpretable linear probabilistic classification with L2 regularization.
4. **Random Forest Classifier**: Non-linear tree ensemble capturing feature interactions and non-linear boundaries.
5. **PyTorch LSTM Sequence Classifier**: Deep recurrent neural network modeling sequence history across 60-minute windows.


### The Question We Are Answering

**In plain language:** _Can a computer learn to predict whether a cryptocurrency's price will go up or down in the next 15 minutes, using only publicly available trading data?_

Every minute, crypto exchanges like Binance publish a data record (called a "k-line") that summarizes what happened in that minute: the opening price, closing price, highest price, lowest price, how much was traded, and how many trades occurred. We have collected **3 years** of this data for the **20 most actively traded** cryptocurrency pairs — that's over **30 million** data points.

This notebook takes that raw data and:
1. **Extracts meaningful signals** — 16 carefully chosen measurements that capture price trends, momentum, volatility, and trading activity patterns.
2. **Trains 5 different prediction methods** — from simple baselines ("always guess the majority direction") to advanced deep learning (LSTM neural networks that read 60-minute sequences).
3. **Compares them fairly** — using strict time-based splitting so no model ever "peeks" at future data during training.

The goal is to determine whether machine learning can do better than traditional statistical methods (and simple guessing) at this prediction task, and if so, by how much.


### Pipeline Architecture & Technical Workflow

```mermaid
graph TD
  A["Raw K-Line Parquet Sample"] --> B["Per-Symbol Feature Calculation"]
  B --> C1["Price Momentum & Trend<br>(log_return, SMA 15/50, EMA 15/50)"]
  B --> C2["Oscillators & Envelopes<br>(RSI-14, MACD Norm, Bollinger %B)"]
  B --> C3["Order Flow & Volatility<br>(taker_buy_ratio, volume_z30, trades_z30, vol_30)"]
  B --> C4["Cyclical Time Encoders<br>(hour_sin, hour_cos)"]
  C1 --> D["16-Feature Canonical Matrix"]
  C2 --> D
  C3 --> D
  C4 --> D
  D --> E["Chronological Data Split & Purging<br>(70% Train / 15% Val / 15% Test)"]
  E --> F1["1. Majority Class Floor Baseline"]
  E --> F2["2. Continuous OLS Return Regression"]
  E --> F3["3. Logistic Regression Classifier (L2)"]
  E --> F4["4. Random Forest Classifier (100 Trees)"]
  E --> F5["5. PyTorch LSTM Sequence Classifier (60-min Lookback)"]
  F3 --> G["Validation Decision Threshold Tuning<br>(Scans cutoffs 0.40–0.60 for Balanced Acc)"]
  F4 --> G
  F5 --> G
  G --> H["Persisted Model Artifacts<br>(models/lstm_checkpoint.pt & models/ml_artifacts.pkl)"]
```


In [1]:
import logging
import os
import sys

import numpy as np
import polars as pl

sys.path.insert(0, os.path.abspath(os.path.join("..", "src")))
sys.path.insert(0, os.path.abspath(os.path.join("..")))

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
)

import src.config as config  # noqa: E402
from src.features.indicators import (  # noqa: E402
    compute_flow_features,
    compute_indicators,
    compute_stationary_features,
)
from src.features.labels import add_direction_labels  # noqa: E402
from src.models.baselines import (  # noqa: E402
    load_baseline_artifacts,
    predict_majority_class,
    predict_ols_direction,
    save_baseline_artifacts,
    train_baselines,
)
from src.models.evaluation import (  # noqa: E402
    calculate_metrics,
    log_metrics,
    save_metrics_json,
)
from src.models.lstm import (  # noqa: E402
    SequenceDataset,
    load_lstm_artifacts,
    predict_lstm,
    save_lstm_artifacts,
    train_lstm,
)
from src.models.train import (  # noqa: E402
    load_model_artifacts,
    prepare_features_and_targets,
    save_model_artifacts,
    split_data_chronologically,
    train_pipeline,
)
from src.pipeline.preprocess import compute_split_boundaries  # noqa: E402

# Feature columns from canonical config
feature_cols = config.FEATURE_COLS

# Fast local execution filter (Set DEV_SYMBOLS = None for full cloud/sample run)
# LOCAL FAST RUN: Filters to 2 symbols for quick iteration.
# For AWS full run: set DEV_SYMBOLS = None (all 20 symbols).
DEV_SYMBOLS: list[str] | None = None  # Production run: all trading pairs

# Load raw parquet sample
INPUT_COLS = [
    "symbol",
    "open_time",
    "open",
    "high",
    "low",
    "close",
    "volume",
    "number_of_trades",
    "taker_buy_base_asset_volume",
]
df = config.load_parquet_auto(config.ACTIVE_DATA_PATH, columns=INPUT_COLS)
if DEV_SYMBOLS:
    df = df.filter(pl.col("symbol").is_in(DEV_SYMBOLS))

2026-07-26 11:34:41,861 [INFO] src.config: Loading Parquet from S3: s3://dat204m-binance-bigdata-hub-sg/sample/binance_sample.parquet
2026-07-26 11:34:41,961 [INFO] botocore.credentials: Found credentials from IAM Role: BaseNotebookInstanceEc2InstanceRole


## 1. Feature Engineering & Target Labelling

### Overview of Model Features (Plain-Language Guide)

To predict cryptocurrency price direction over the next 15 minutes, the pipeline constructs **16 quantitative features** categorized into five intuitive domain areas:

| Feature | Category | Plain-Language Description | Financial & Domain Intuition |
|:--- |:--- |:--- |:--- |
| `log_return` | Price Dynamics | 1-minute percentage price change (log scale) | Measures immediate price momentum without compounding scale distortion. |
| `close_to_sma_15` | Trend Ratio | Ratio of current price to 15-minute simple moving average | Values > 1.0 indicate price is trading above its short-term baseline. |
| `close_to_sma_50` | Trend Ratio | Ratio of current price to 50-minute simple moving average | Identifies medium-term trend direction (bullish vs. bearish alignment). |
| `close_to_ema_15` | Trend Ratio | Ratio of current price to 15-minute exponential moving average | Gives higher weight to recent minutes to react faster to sudden price moves. |
| `close_to_ema_50` | Trend Ratio | Ratio of current price to 50-minute exponential moving average | Tracks medium-term momentum with recency weighting. |
| `bb_position` | Volatility Envelope | Relative price position within 20-minute Bollinger Bands (0.0 to 1.0) | 1.0 = upper band (potentially overbought), 0.0 = lower band (oversold). |
| `volatility_30` | Volatility | Standard deviation of 1-minute returns over past 30 minutes | Quantifies current market turbulence and price uncertainty. |
| `rsi_14` | Momentum Speedometer | Relative Strength Index (0–100) over 14 minutes | >70 signals overbought momentum; <30 signals oversold momentum. |
| `macd_line_norm` | Trend Convergence | Short-term EMA minus long-term EMA, normalized by price | Tracks overall trend speed and direction shifts. |
| `macd_signal_norm` | Signal Baseline | 9-minute moving average of MACD line | Trigger line: MACD crossing above Signal indicates accelerating upward trend. |
| `macd_hist_norm` | Momentum Barometer | Difference between MACD line and Signal line | Positive values reflect growing bullish momentum strength. |
| `taker_buy_ratio` | Order Flow Pressure | Ratio of buyer-initiated volume to total volume (0.0 to 1.0) | >0.5 means aggressive buyers taking liquidity; <0.5 means sellers dominate. |
| `volume_z30` | Volume Anomaly | Current trading volume standardized relative to past 30-min average | Identifies unusual volume spikes (e.g. +2.0 = volume is 2 std devs above normal). |
| `trades_z30` | Trade Frequency | Number of trades standardized relative to past 30-min average | High trade activity signals institutional participation or sudden market interest. |
| `hour_sin` | Time Cycle | Sine component of 24-hour trading cycle | Encodes time of day continuously so 23:59 and 00:00 remain adjacent. |
| `hour_cos` | Time Cycle | Cosine component of 24-hour trading cycle | Works with `hour_sin` to capture global market open/close sessions (Asia, Europe, US). |

---


In [2]:
print("Generating technical indicators, flow features, and stationary features (per symbol)...")

# Memory: process one symbol at a time and keep only the modeling columns as
# float32, instead of materializing the full wide frame for all symbols.
# Features are computed per symbol anyway, so the result is identical.
KEEP_COLS = ["symbol", "open_time", "close", *feature_cols]

symbol_frames = []
for symbol in df["symbol"].unique().sort().to_list():
    part = df.filter(pl.col("symbol") == symbol)
    part = compute_stationary_features(compute_indicators(part))
    part = compute_flow_features(part)
    part = part.select(KEEP_COLS).with_columns([pl.col(c).cast(pl.Float32) for c in feature_cols])
    symbol_frames.append(part)

df_features = pl.concat(symbol_frames)
del df, symbol_frames, part  # free memory (re-run from the top if re-executing)

# Binary direction label: 1 if close[t+H] exceeds close[t] by more than the threshold
df_labeled = add_direction_labels(
    df_features, horizon=config.FUTURE_HORIZON, threshold=config.TARGET_THRESHOLD
)
del df_features

# Drop warm-up nulls (rolling windows) and unlabeled tail rows (no future close)
df_labeled = df_labeled.drop_nulls()
print(f"Dataset ready for training. Total clean rows: {len(df_labeled):,}")

Generating technical indicators, flow features, and stationary features (per symbol)...
Dataset ready for training. Total clean rows: 30,657,940


## 2. Chronological Data Splitting & Purging

### Why Chronological Splitting?
In financial time series, random train/test splitting (k-fold cross-validation) is illegal because future information leaks into past predictions. We enforce a strict timeline split:
- **Training Set (70%)**: Used to fit model parameters.
- **Validation Set (15%)**: Used to tune decision thresholds and hyperparameter selection.
- **Test Set (15%)**: Held-out partition evaluated strictly in Notebook 03.

### Label Purging
Because target labels look ahead 15 minutes (`FUTURE_HORIZON=15`), samples near boundary dates overlap in time. We purge 15 minutes of data immediately preceding partition boundaries to ensure complete independence between training, validation, and testing sets.


In [3]:
# Boundary timestamps at the 70th/85th percentile of the global timeline
train_end, val_end = compute_split_boundaries(df_labeled, train_frac=0.70, val_frac=0.15)
print(f"Train until {train_end}, validate until {val_end}, test afterwards")

# Purge = FUTURE_HORIZON minutes: drops rows whose label window crosses a boundary
train_df, val_df, test_df = split_data_chronologically(
    df_labeled, train_end, val_end, purge_minutes=config.FUTURE_HORIZON
)
del df_labeled, test_df  # test partition is rebuilt and used in notebook 03

2026-07-26 11:35:00,792 [INFO] src.models.train: Chronologically splitting data: train_end=2025-06-29 19:52:00, val_end=2025-12-14 21:48:00, purge_minutes=15


Train until 2025-06-29 19:52:00, validate until 2025-12-14 21:48:00, test afterwards


2026-07-26 11:35:01,005 [INFO] src.models.train: Split sizes - Train: 21460553, Val: 4598399, Test: 4598418


## 3. Feature & Target Matrices

All models use the same 16 features (11 stationary price features + 5 order-flow/time features). The classifiers use the binary `target`; the OLS baseline uses the continuous `future_return` from the same rows.

In [4]:
# Extract feature/target matrices (feature_cols defined at the top of the notebook)
X_train, y_train = prepare_features_and_targets(train_df, feature_cols, "target")
X_val, y_val = prepare_features_and_targets(val_df, feature_cols, "target")

# Continuous future returns for the OLS baseline (same rows: target is null iff future_return is)
_, r_train = prepare_features_and_targets(train_df, feature_cols, "future_return")
# train_df, val_df preserved for PyTorch LSTM sequence generation

print(f"Train matrix: {X_train.shape}, Val matrix: {X_val.shape}")
print(f"Train class balance (share of 'up'): {y_train.mean():.4f}")

2026-07-26 11:35:01,229 [INFO] src.models.train: Feature matrix shape: (21460553, 16), target 'target' positive rate: 0.4815
2026-07-26 11:35:01,255 [INFO] src.models.train: Feature matrix shape: (4598399, 16), target 'target' positive rate: 0.4761
2026-07-26 11:35:01,539 [INFO] src.models.train: Feature matrix shape: (21460553, 16), target 'future_return' positive rate: 0.0000


Train matrix: (21460553, 16), Val matrix: (4598399, 16)
Train class balance (share of 'up'): 0.4815


## 4. Baselines

- Majority class: always predicts the most common training label. Every model must beat this.
- OLS: linear regression predicts the 15-min future return; prediction above the threshold means up. Fit on the training partition only.

In [5]:
# Fit both baselines on the training partition only (or load existing checkpoint)
baseline_path = config.PROJECT_ROOT / "models" / "baseline_artifacts.pkl"
if baseline_path.exists():
    print(f"Found existing baseline artifacts at {baseline_path}. Loading without retraining...")
    baselines = load_baseline_artifacts(baseline_path)
else:
    print("Fitting baseline models (Majority Class & OLS)...")
    baselines = train_baselines(
        X_train, y_train, r_train, feature_cols, threshold=config.TARGET_THRESHOLD
    )
    save_baseline_artifacts(baselines, config.PROJECT_ROOT / "models")

# Evaluate on the validation partition
majority_val_preds = predict_majority_class(baselines.majority_label, len(y_val))
ols_val_preds = predict_ols_direction(baselines.ols_model, X_val, baselines.decision_threshold)

baseline_val_metrics = {
    "majority_class": calculate_metrics(y_val, majority_val_preds),
    "ols_threshold": calculate_metrics(y_val, ols_val_preds),
}
for name, metrics in baseline_val_metrics.items():
    log_metrics(metrics, model_name=name)

Found existing baseline artifacts at /home/ec2-user/SageMaker/dat204m-final-project/models/baseline_artifacts.pkl. Loading without retraining...


2026-07-26 11:35:10,173 [INFO] src.models.evaluation: === Evaluation Metrics for majority_class ===
2026-07-26 11:35:10,174 [INFO] src.models.evaluation: Acc: 0.5239 | Prec: 0.0000 | Rec: 0.0000 | F1: 0.0000 | Bal Acc: 0.5000
2026-07-26 11:35:10,174 [INFO] src.models.evaluation: Confusion Matrix: TN=2408975, FP=0, FN=2189424, TP=0
2026-07-26 11:35:10,175 [INFO] src.models.evaluation: === Evaluation Metrics for ols_threshold ===
2026-07-26 11:35:10,175 [INFO] src.models.evaluation: Acc: 0.5069 | Prec: 0.4810 | Rec: 0.4506 | F1: 0.4653 | Bal Acc: 0.5044
2026-07-26 11:35:10,175 [INFO] src.models.evaluation: Confusion Matrix: TN=1344545, FP=1064430, FN=1202880, TP=986544


## 5. Logistic Regression & Random Forest

Logistic Regression is the linear ML model (interpretable coefficients). Random Forest is the nonlinear one (feature importances). Validation metrics for tabular models are printed below; full method ladder evaluation including LSTM is presented after section 6.

In [6]:
# Fit scaler + Logistic Regression + Random Forest (or load existing checkpoint)
ml_artifacts_path = config.PROJECT_ROOT / "models" / "ml_artifacts.pkl"
if ml_artifacts_path.exists():
    print(f"Found existing ML artifacts at {ml_artifacts_path}. Loading without retraining...")
    artifacts = load_model_artifacts(ml_artifacts_path)
else:
    print("Training Logistic Regression and Random Forest models...")
    artifacts = train_pipeline(X_train, y_train, X_val, y_val, feature_cols)
    save_model_artifacts(artifacts, config.PROJECT_ROOT / "models")

# Validation comparison across tabular models
X_val_scaled = artifacts.scaler.transform(X_val)
lr_val_preds = artifacts.logistic_regression.predict(X_val_scaled)
rf_val_preds = artifacts.random_forest.predict(X_val_scaled)

comparison = {
    "Majority class": baseline_val_metrics["majority_class"],
    "OLS -> threshold": baseline_val_metrics["ols_threshold"],
    "Logistic Regression": calculate_metrics(y_val, lr_val_preds),
    "Random Forest": calculate_metrics(y_val, rf_val_preds),
}
comparison_df = pl.DataFrame([{"model": name, **m} for name, m in comparison.items()])
print(comparison_df)

2026-07-26 11:35:10,182 [INFO] src.models.train: Loading ML artifacts from /home/ec2-user/SageMaker/dat204m-final-project/models/ml_artifacts.pkl...


Found existing ML artifacts at /home/ec2-user/SageMaker/dat204m-final-project/models/ml_artifacts.pkl. Loading without retraining...
shape: (4, 9)
┌────────────┬──────────┬───────────┬──────────┬───┬────────────┬───────────┬──────────┬───────────┐
│ model      ┆ accuracy ┆ precision ┆ recall   ┆ … ┆ balanced_a ┆ specifici ┆ npv      ┆ confusion │
│ ---        ┆ ---      ┆ ---       ┆ ---      ┆   ┆ ccuracy    ┆ ty        ┆ ---      ┆ _matrix   │
│ str        ┆ f64      ┆ f64       ┆ f64      ┆   ┆ ---        ┆ ---       ┆ f64      ┆ ---       │
│            ┆          ┆           ┆          ┆   ┆ f64        ┆ f64       ┆          ┆ struct[4] │
╞════════════╪══════════╪═══════════╪══════════╪═══╪════════════╪═══════════╪══════════╪═══════════╡
│ Majority   ┆ 0.523873 ┆ 0.0       ┆ 0.0      ┆ … ┆ 0.5        ┆ 1.0       ┆ 0.523873 ┆ {2408975, │
│ class      ┆          ┆           ┆          ┆   ┆            ┆           ┆          ┆ 0,2189424 │
│            ┆          ┆           ┆        

## 6. PyTorch LSTM Sequence Classifier & Hyperparameter Sensitivity

The PyTorch LSTM operates on sequence windows `(batch, seq_len, 16)` constructed per symbol via `SequenceDataset` to prevent cross-symbol boundary contamination. We train a baseline 2-layer LSTM and run a 5-configuration hyperparameter sweep to test context length (`seq_len`), model capacity (`hidden_size`), and regularization (`dropout`).


In [7]:
# Section 5.1: Train PyTorch LSTM Sequence Classifier (Baseline Config A)
cfg_a_path = config.PROJECT_ROOT / "models" / "lstm_checkpoint_config_a_baseline.pt"
main_lstm_path = config.PROJECT_ROOT / "models" / "lstm_checkpoint.pt"

if cfg_a_path.exists():
    print(f"Found existing checkpoint for Config A at {cfg_a_path}. Loading without retraining...")
    loaded_a = load_lstm_artifacts(cfg_a_path)
    lstm_model_a, lstm_scaler_a, lstm_thr_a, lstm_hist_a = (
        loaded_a["model"],
        loaded_a["scaler"],
        loaded_a["threshold"],
        loaded_a["history"],
    )
elif main_lstm_path.exists():
    print(f"Found existing main LSTM checkpoint at {main_lstm_path}. Loading without retraining...")
    loaded_a = load_lstm_artifacts(main_lstm_path)
    lstm_model_a, lstm_scaler_a, lstm_thr_a, lstm_hist_a = (
        loaded_a["model"],
        loaded_a["scaler"],
        loaded_a["threshold"],
        loaded_a["history"],
    )
else:
    print("--- Training PyTorch LSTM Baseline (Config A: seq_len=60, hidden=64, dropout=0.3) ---")
    lstm_model_a, lstm_scaler_a, lstm_thr_a, lstm_hist_a = train_lstm(
        train_df=train_df,
        val_df=val_df,
        feature_cols=feature_cols,
        target_col="target",
        seq_len=60,
        hidden_size=64,
        dropout=0.3,
        max_epochs=20,  # LOCAL: set to 20+ for AWS full run
        patience=3,
        config_name="Config A (baseline)",
        log_filepath=config.PROJECT_ROOT / "models" / "lstm_training_log.jsonl",
    )
    save_lstm_artifacts(
        model=lstm_model_a,
        scaler=lstm_scaler_a,
        threshold=lstm_thr_a,
        feature_cols=feature_cols,
        seq_len=60,
        hparams={"name": "Config A (baseline)", "seq_len": 60, "hidden_size": 64, "dropout": 0.3},
        filepath=cfg_a_path,
        history=lstm_hist_a,
    )

print(
    f"Config A Baseline Val Balanced Acc: {lstm_hist_a.get('best_val_balanced_acc', 0.0):.4f} @ threshold {lstm_thr_a:.3f}"
)

Found existing checkpoint for Config A at /home/ec2-user/SageMaker/dat204m-final-project/models/lstm_checkpoint_config_a_baseline.pt. Loading without retraining...
Config A Baseline Val Balanced Acc: 0.5244 @ threshold 0.475


### 6.2 LSTM Hyperparameter Sensitivity Sweep

We evaluate 5 representative candidate configurations on the validation partition:
- **Config A (Baseline)**: `seq_len=60`, `hidden_size=64`, `dropout=0.3` (1-hour lookback, moderate capacity)
- **Config B (Short Context)**: `seq_len=30`, `hidden_size=64`, `dropout=0.3` (30-min lookback)
- **Config C (Long Context)**: `seq_len=120`, `hidden_size=64`, `dropout=0.3` (2-hour lookback)
- **Config D (Higher Capacity)**: `seq_len=60`, `hidden_size=128`, `dropout=0.3` (128 hidden units)
- **Config E (Higher Dropout)**: `seq_len=60`, `hidden_size=64`, `dropout=0.5` (heavy regularization)


In [8]:
# Sweep execution across candidates A-E
configs = [
    {"name": "Config A (baseline)", "seq_len": 60, "hidden_size": 64, "dropout": 0.3},
    {"name": "Config B (seq_len=30)", "seq_len": 30, "hidden_size": 64, "dropout": 0.3},
    {"name": "Config C (seq_len=120)", "seq_len": 120, "hidden_size": 64, "dropout": 0.3},
    {"name": "Config D (hidden=128)", "seq_len": 60, "hidden_size": 128, "dropout": 0.3},
    {"name": "Config E (dropout=0.5)", "seq_len": 60, "hidden_size": 64, "dropout": 0.5},
]

sweep_results = []
best_overall_acc = -1.0
winning_artifacts = None

main_lstm_path = config.PROJECT_ROOT / "models" / "lstm_checkpoint.pt"

for cfg in configs:
    slug = cfg["name"].lower().replace(" ", "_").replace("(", "").replace(")", "").replace("=", "_")
    cfg_path = config.PROJECT_ROOT / "models" / f"lstm_checkpoint_{slug}.pt"

    if cfg_path.exists():
        print(
            f"\nFound existing checkpoint for {cfg['name']} at {cfg_path}. Loading without retraining..."
        )
        loaded = load_lstm_artifacts(cfg_path)
        model, scaler, thr, hist = (
            loaded["model"],
            loaded["scaler"],
            loaded["threshold"],
            loaded["history"],
        )
    elif cfg["name"] == "Config A (baseline)" and main_lstm_path.exists():
        print(
            f"\nFound existing main LSTM checkpoint at {main_lstm_path} for {cfg['name']}. Loading without retraining..."
        )
        loaded = load_lstm_artifacts(main_lstm_path)
        model, scaler, thr, hist = (
            loaded["model"],
            loaded["scaler"],
            loaded["threshold"],
            loaded["history"],
        )
    else:
        print(f"\nEvaluating {cfg['name']}...")
        if cfg["name"] == "Config A (baseline)":
            model, scaler, thr, hist = lstm_model_a, lstm_scaler_a, lstm_thr_a, lstm_hist_a
        else:
            model, scaler, thr, hist = train_lstm(
                train_df=train_df,
                val_df=val_df,
                feature_cols=feature_cols,
                target_col="target",
                seq_len=cfg["seq_len"],
                hidden_size=cfg["hidden_size"],
                dropout=cfg["dropout"],
                max_epochs=20,  # LOCAL: set to 20+ for AWS full run
                patience=3,
                config_name=cfg["name"],
                log_filepath=config.PROJECT_ROOT / "models" / "lstm_training_log.jsonl",
            )
        save_lstm_artifacts(
            model=model,
            scaler=scaler,
            threshold=thr,
            feature_cols=feature_cols,
            seq_len=cfg["seq_len"],
            hparams=cfg,
            filepath=cfg_path,
            history=hist,
        )

    bal_acc = hist.get("best_val_balanced_acc", 0.0)
    sweep_results.append(
        {
            "Config": cfg["name"],
            "Seq Len": cfg["seq_len"],
            "Hidden Size": cfg["hidden_size"],
            "Dropout": cfg["dropout"],
            "Val Balanced Acc": bal_acc,
            "Tuned Threshold": thr,
        }
    )
    if bal_acc > best_overall_acc:
        best_overall_acc = bal_acc
        winning_artifacts = (model, scaler, thr, cfg, hist)

sweep_df = pl.DataFrame(
    [
        {
            "Config": r["Config"],
            "Seq Len": r["Seq Len"],
            "Hidden Size": r["Hidden Size"],
            "Dropout": r["Dropout"],
            "Val Balanced Acc": r["Val Balanced Acc"],
            "Tuned Threshold": r["Tuned Threshold"],
        }
        for r in sweep_results
    ]
)
print("\n=== LSTM Hyperparameter Sensitivity Sweep Results ===")
print(sweep_df)

# Extract winning artifacts
lstm_model, lstm_scaler, lstm_best_threshold, lstm_winning_cfg, lstm_hist = winning_artifacts
if "val_probs" in lstm_hist and "val_targets" in lstm_hist:
    lstm_val_probs = lstm_hist["val_probs"]
    lstm_val_targets = lstm_hist["val_targets"]
else:
    val_ds = SequenceDataset(
        val_df,
        feature_cols=feature_cols,
        target_col="target",
        seq_len=lstm_winning_cfg["seq_len"],
        scaler=lstm_scaler,
    )
    lstm_val_probs, _ = predict_lstm(lstm_model, val_ds)
    lstm_val_targets = val_ds.targets[val_ds.valid_end_indices]
lstm_val_preds_default = (lstm_val_probs >= 0.5).astype(int)
lstm_val_metrics_default = calculate_metrics(
    lstm_val_targets, lstm_val_preds_default, lstm_val_probs
)

# Validation comparison across the full method ladder
full_comparison = {
    "Majority class": baseline_val_metrics["majority_class"],
    "OLS -> threshold": baseline_val_metrics["ols_threshold"],
    "Logistic Regression": calculate_metrics(y_val, lr_val_preds),
    "Random Forest": calculate_metrics(y_val, rf_val_preds),
    "PyTorch LSTM": lstm_val_metrics_default,
}
full_comparison_df = pl.DataFrame([{"model": name, **m} for name, m in full_comparison.items()])
print("\n=== Validation Comparison Across Full Method Ladder ===")
print(full_comparison_df)

# Save winning checkpoint
save_lstm_artifacts(
    model=lstm_model,
    scaler=lstm_scaler,
    threshold=lstm_best_threshold,
    feature_cols=feature_cols,
    seq_len=lstm_winning_cfg["seq_len"],
    hparams=lstm_winning_cfg,
    filepath=config.PROJECT_ROOT / "models" / "lstm_checkpoint.pt",
    history=lstm_hist,
    metrics=lstm_val_metrics_default,
)
print(f"\nWinning Config: {lstm_winning_cfg['name']} saved to models/lstm_checkpoint.pt")

# Export sweep metrics to JSON
sweep_export = {
    r["Config"]: {
        "hparams": {
            "seq_len": r["Seq Len"],
            "hidden_size": r["Hidden Size"],
            "dropout": r["Dropout"],
        },
        "best_threshold": r["Tuned Threshold"],
        "best_val_balanced_acc": r["Val Balanced Acc"],
    }
    for r in sweep_results
}
save_metrics_json(
    sweep_export,
    config.PROJECT_ROOT / "data" / "sample" / "lstm_sweep_metrics.json",
)
# Free training/validation polars DataFrames after LSTM sequence creation


Found existing checkpoint for Config A (baseline) at /home/ec2-user/SageMaker/dat204m-final-project/models/lstm_checkpoint_config_a_baseline.pt. Loading without retraining...

Found existing checkpoint for Config B (seq_len=30) at /home/ec2-user/SageMaker/dat204m-final-project/models/lstm_checkpoint_config_b_seq_len_30.pt. Loading without retraining...

Found existing checkpoint for Config C (seq_len=120) at /home/ec2-user/SageMaker/dat204m-final-project/models/lstm_checkpoint_config_c_seq_len_120.pt. Loading without retraining...

Found existing checkpoint for Config D (hidden=128) at /home/ec2-user/SageMaker/dat204m-final-project/models/lstm_checkpoint_config_d_hidden_128.pt. Loading without retraining...

Found existing checkpoint for Config E (dropout=0.5) at /home/ec2-user/SageMaker/dat204m-final-project/models/lstm_checkpoint_config_e_dropout_0.5.pt. Loading without retraining...

=== LSTM Hyperparameter Sensitivity Sweep Results ===
shape: (5, 6)
┌────────────────────────┬─────

2026-07-26 11:35:48,124 [INFO] src.models.lstm: Saved LSTM artifacts to /home/ec2-user/SageMaker/dat204m-final-project/models/lstm_checkpoint.pt
2026-07-26 11:35:48,125 [INFO] src.models.evaluation: Saved evaluation metrics JSON to /home/ec2-user/SageMaker/dat204m-final-project/models/lstm_checkpoint_metrics.json
2026-07-26 11:35:48,127 [INFO] src.models.evaluation: Saved evaluation metrics JSON to /home/ec2-user/SageMaker/dat204m-final-project/data/sample/lstm_sweep_metrics.json



Winning Config: Config E (dropout=0.5) saved to models/lstm_checkpoint.pt


## 7. Decision Threshold Tuning

The classifiers output a probability of "up", and a cutoff turns that into a prediction. The default cutoff is 0.5, but our training data is only ~48% "up", so predicted probabilities cluster just below 0.5. With the default cutoff the models almost never predict "up" — they mostly learned to say "down".

To fix this we select the cutoff on the **validation set** (never the test set): we try cutoffs from 0.40 to 0.60 in steps of 0.005 and keep the one with the best **balanced accuracy** (the average of accuracy on "up" rows and accuracy on "down" rows — a metric that rewards getting both classes right instead of siding with the majority).

The exact numbers used are printed below and saved inside the model artifacts, so notebook 03 applies the same cutoffs on the test set. We report both the default and tuned versions there: the tuned cutoff trades a little raw accuracy for much better recall and balanced accuracy.

Justification summary:
- Why not 0.5: it sits above most predicted probabilities due to class imbalance, collapsing "up" recall.
- Why chosen on validation: choosing it on test would leak the answer key into the decision.
- Why balanced accuracy: plain accuracy is maximized by predicting the majority class everywhere; balanced accuracy is not.

In [9]:
from sklearn.metrics import balanced_accuracy_score

# Compute validation probabilities and tune cutoffs for max balanced accuracy
X_val_scaled = artifacts.scaler.transform(X_val)
lr_val_probs = artifacts.logistic_regression.predict_proba(X_val_scaled)[:, 1]
rf_val_probs = artifacts.random_forest.predict_proba(X_val_scaled)[:, 1]


def find_best_threshold(y_true, y_probs):
    best_thr, best_score = 0.5, -1.0
    for thr in np.arange(0.40, 0.605, 0.005):
        score = balanced_accuracy_score(y_true, (y_probs >= thr).astype(int))
        if score > best_score:
            best_score, best_thr = score, float(thr)
    return best_thr


artifacts.lr_threshold = find_best_threshold(y_val, lr_val_probs)
artifacts.rf_threshold = find_best_threshold(y_val, rf_val_probs)
save_model_artifacts(artifacts, config.PROJECT_ROOT / "models")

# The exact thresholds selected on the validation set (saved with the model artifacts)
print(f"Train class balance (share of 'up'): {y_train.mean():.4f}")
print(f"Logistic Regression decision threshold: 0.500 -> {artifacts.lr_threshold:.3f}")
print(f"Random Forest decision threshold:    0.500 -> {artifacts.rf_threshold:.3f}")
print(f"PyTorch LSTM decision threshold:     0.500 -> {lstm_best_threshold:.3f}")

threshold_rows = []
for name, probs, tuned_thr, y_true in [
    ("Logistic Regression", lr_val_probs, artifacts.lr_threshold, y_val),
    ("Random Forest", rf_val_probs, artifacts.rf_threshold, y_val),
    ("PyTorch LSTM", lstm_val_probs, lstm_best_threshold, lstm_val_targets),
]:
    for setting, thr in [("default 0.500", 0.5), (f"tuned {tuned_thr:.3f}", tuned_thr)]:
        metrics = calculate_metrics(y_true, (probs >= thr).astype(int), probs)
        threshold_rows.append({"model": name, "threshold": setting, **metrics})

threshold_df = pl.DataFrame(threshold_rows)
print(threshold_df)

2026-07-26 11:36:54,891 [INFO] src.models.train: Saving ML artifacts to /home/ec2-user/SageMaker/dat204m-final-project/models/ml_artifacts.pkl...
2026-07-26 11:36:54,928 [INFO] src.models.evaluation: Saved evaluation metrics JSON to /home/ec2-user/SageMaker/dat204m-final-project/models/ml_metrics.json
2026-07-26 11:36:54,929 [INFO] src.models.train: Model artifacts saved successfully.


Train class balance (share of 'up'): 0.4815
Logistic Regression decision threshold: 0.500 -> 0.480
Random Forest decision threshold:    0.500 -> 0.480
PyTorch LSTM decision threshold:     0.500 -> 0.475
shape: (6, 13)
┌────────────┬────────────┬──────────┬───────────┬───┬────────────┬──────────┬──────────┬──────────┐
│ model      ┆ threshold  ┆ accuracy ┆ precision ┆ … ┆ confusion_ ┆ roc_auc  ┆ pr_auc   ┆ log_loss │
│ ---        ┆ ---        ┆ ---      ┆ ---       ┆   ┆ matrix     ┆ ---      ┆ ---      ┆ ---      │
│ str        ┆ str        ┆ f64      ┆ f64       ┆   ┆ ---        ┆ f64      ┆ f64      ┆ f64      │
│            ┆            ┆          ┆           ┆   ┆ struct[4]  ┆          ┆          ┆          │
╞════════════╪════════════╪══════════╪═══════════╪═══╪════════════╪══════════╪══════════╪══════════╡
│ Logistic   ┆ default    ┆ 0.525237 ┆ 0.503687  ┆ … ┆ {1986671,4 ┆ 0.522942 ┆ 0.495205 ┆ 0.691442 │
│ Regression ┆ 0.500      ┆          ┆           ┆   ┆ 22304,1760 ┆        

## 8. Hyperparameters & Overfitting Check

Hyperparameters used (defined in `src/models/train.py`):

| Model | Hyperparameters | Reason |
|---|---|---|
| Logistic Regression | C=0.1 (L2 regularization), max_iter=1000 | Stronger regularization guards against instability from the correlated features |
| Random Forest | n_estimators=100, max_depth=10, n_jobs=2 | Depth cap limits overfitting; workers bounded to 2 to keep peak memory within a 16 GB machine |
| OLS baseline | none (closed-form least squares) | — |
| PyTorch LSTM | 2 layers, hidden=64, dropout=0.3, AdamW lr=1e-3 | Winning config from 5-candidate sweep |
| Decision threshold | scanned 0.40–0.60 in steps of 0.005 on validation | see Section 6 |

The cell below compares train vs validation accuracy. A large gap (train much higher) would indicate overfitting; near-equal values indicate the models generalize.


In [10]:
# Overfitting check: compare train vs validation accuracy for both classifiers
X_train_scaled = artifacts.scaler.transform(X_train)
lr_train_acc = artifacts.logistic_regression.score(X_train_scaled, y_train)
rf_train_acc = artifacts.random_forest.score(X_train_scaled, y_train)
del X_train_scaled  # free the 21M-row scaled copy

lr_val_acc = artifacts.logistic_regression.score(X_val_scaled, y_val)
rf_val_acc = artifacts.random_forest.score(X_val_scaled, y_val)

print(f"Logistic Regression - train: {lr_train_acc:.4f}, val: {lr_val_acc:.4f}")
print(f"Random Forest    - train: {rf_train_acc:.4f}, val: {rf_val_acc:.4f}")
print("A small train-val gap means the models are not overfitting.")

hist_dict = lstm_hist.get("history", lstm_hist)
if "train_acc" in hist_dict:
    lstm_train_acc = hist_dict["train_acc"][-1]
elif "train_loss" in hist_dict and len(hist_dict["train_loss"]) > 0:
    lstm_train_acc = 1.0 - hist_dict["train_loss"][-1]
else:
    lstm_train_acc = 0.0
print(
    f"PyTorch LSTM    - val acc (default): {lstm_val_metrics_default['accuracy']:.4f}, val bal acc (tuned): {best_overall_acc:.4f}"
)

Logistic Regression - train: 0.5263, val: 0.5252
Random Forest    - train: 0.5311, val: 0.5273
A small train-val gap means the models are not overfitting.
PyTorch LSTM    - val acc (default): 0.5299, val bal acc (tuned): 0.5245


## What the Training Results Tell Us

After running all 5 methods on the training data and checking them on the validation set:

- **The simple baseline** (always guess "down") achieves **52.39%** validation accuracy, reflecting slight market downward bias during the validation period.
- **Traditional statistics** (OLS linear regression) underperforms at **50.69%** validation accuracy — essentially no predictive power.
- **Tabular machine learning works** — both Logistic Regression (**52.52%** accuracy) and Random Forest (**52.73%** accuracy) beat the baseline on validation data.
- **The PyTorch LSTM neural network** achieves the highest validation performance — **52.99%** default accuracy and **52.45%** tuned balanced accuracy (0.5346 ROC-AUC).

The edge is small (~0.6% over the baseline on validation) but consistent across millions of data points. **Notebook 03** evaluates whether this edge holds up on the completely held-out test set.

## References

1. Binance Public Data — historical spot 1-minute k-lines. https://github.com/binance/binance-public-data
2. Pedregosa et al. (2011). Scikit-learn: Machine Learning in Python. JMLR 12, 2825-2830.
3. Wilder, J. W. (1978). *New Concepts in Technical Trading Systems* — RSI definition.
4. Bollinger, J. (2001). *Bollinger on Bollinger Bands* — Bollinger Bands definition.
5. López de Prado, M. (2018). *Advances in Financial Machine Learning* — chronological splitting and purging for financial time series.